source/ice_api.c

```diff
@@ -0,0 +1,1676 @@
+            pIpAddr,
+            sizeof( IceIPAddress_t ) );
+    pCandidate->state = ICE_CANDIDATE_STATE_VALID;
+
+    for( i = 0; ( ( i < Ice_GetValidRemoteCandidateCount( pIceAgent ) ) && ( retStatus == ICE_RESULT_OK ) ); i++ )
+    {
+        retStatus = Ice_CreateCandidatePair( pIceAgent,
+                                             pCandidate,
+                                             &( pIceAgent->remoteCandidates[ i ] ) );
+    }
+
+    return retStatus;
+}
+
+/*------------------------------------------------------------------------------------------------------------------*/
+
+/* Ice_InitializeStunPacket - This API populates the Stun packet, whose memory has been allocated by the application.
+ *  4 types of packets need to be created:
+ *   1. Send Srflx Request
+ *   2. Connectivity Check
+ *   3. During nomination - USE_CANDIDATE flag
+ *   4. Send Response to Remote Candidates
+ */
+IceResult_t Ice_InitializeStunPacket( IceAgent_t * pIceAgent,
+                                      StunContext_t * pStunCxt,
+                                      uint8_t * pTransactionId,
+                                      uint8_t * pStunMessageBuffer,
+                                      StunHeader_t * pStunHeader,
+                                      uint8_t isGenerateTransactionID,
+                                      uint8_t isStunBindingRequest )
+{
+    IceResult_t retStatus = ICE_RESULT_OK;
+    StunResult_t stunRetStatus = STUN_RESULT_OK;
+    int i;
+
+    if( ( pIceAgent == NULL ) ||
+        ( pStunCxt == NULL ) ||
+        ( pTransactionId == NULL ) ||
+        ( pStunMessageBuffer == NULL ) ||
+        ( pStunHeader == NULL ) )
+    {
+        retStatus = ICE_RESULT_BAD_PARAM;
+    }
+
+    pStunHeader->pTransactionId = pTransactionId;
+
+    /* STUN header */
+    if( retStatus == ICE_RESULT_OK )
+    {
+        if( isStunBindingRequest )
+        {
+            pStunHeader->messageType = STUN_MESSAGE_TYPE_BINDING_REQUEST;
+        }
+        else
+        {
+            pStunHeader->messageType = STUN_MESSAGE_TYPE_BINDING_SUCCESS_RESPONSE;
+        }
+
+        if( isGenerateTransactionID == 1 )
+        {
+            for( i = 0; i < STUN_HEADER_TRANSACTION_ID_LENGTH; i++ )
+            {
+                pStunHeader->pTransactionId[ i ] = ( uint8_t ) ( pIceAgent->computeRandom() % 0x100 );
+            }
+        }
+        else if( isGenerateTransactionID == 0 )
+        {
+            memcpy( &( pStunHeader->pTransactionId[ 0 ] ),
+                    &( pTransactionId[ 0 ] ),
+                    STUN_HEADER_TRANSACTION_ID_LENGTH );
+        }
+
+        /* Create a STUN message. */
+        stunRetStatus = StunSerializer_Init( pStunCxt,
+                                             pStunMessageBuffer,
+                                             ICE_STUN_MESSAGE_BUFFER_SIZE,
+                                             pStunHeader );
+    }
+
+    if( stunRetStatus == STUN_RESULT_OK )
```

Comment:
We should have a convert error function to map stun errors to ICE errors, please check [this for reference](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/develop/src/source/Stun/StunUtils.c)

Reply:
If there is only one ICE error - ICE_RESULT_STUN_ERROR mapping to all stun errors then it seems fine.

Reply:
`ICE_RESULT_STUN_ERROR` is the error code mapping to all STUN errors for initializing and packaging the stun packet. For Deserialization of the stun packet , we have handled the error in the following struct
```
typedef enum IceStunPacketHandleResult
{
    ICE_RESULT_STUN_DESERIALIZE_OK,
    ICE_RESULT_STUN_DESERIALIZE_ERROR,
    ICE_RESULT_STUN_INVALID_PACKET_TYPE,
    ICE_RESULT_FOUND_PEER_REFLEXIVE_CANDIDATE,
    ICE_RESULT_UPDATED_SRFLX_CANDIDATE_ADDRESS,
    ICE_RESULT_SEND_RESPONSE_FOR_REMOTE_REQUEST,
    ICE_RESULT_SEND_TRIGGERED_CHECK,
    ICE_RESULT_START_NOMINATION,
    ICE_RESULT_USE_CANDIDATE_FLAG,
    ICE_RESULT_SEND_RESPONSE_FOR_NOMINATION,
    ICE_RESULT_CANDIDATE_PAIR_READY,
    ICE_RESULT_NOT_FOUND_CANDIDATE,
    ICE_RESULT_NOT_FOUND_CANDIDATE_PAIR,
    ICE_RESULT_NOT_FOUND_ADDRESS_ATTRIBUTE,
    ICE_RESULT_NOT_FOUND_MATCHING_TRANSACTION_ID,
    ICE_RESULT_STUN_BINDING_INDICATION
} IceStunPacketHandleResult_t;
```